File: api/crs/codegen/lib/crs_solve_utril/crs_solve_utril.c
#include "crs_solve_utril.h"

void crs_solve_utril(const CRS_Matrix *L, emxArray_real_T *b)
{
  int n;
  int i;
  int b_i;
  int i1;
  int k;
  n = L->row_ptr->size[0];
  for (i = 0; i <= n - 2; i++) {
    b_i = L->row_ptr->data[i];
    i1 = L->row_ptr->data[i + 1] - 1;
    for (k = b_i; k <= i1; k++) {
      b->data[i] -= L->val->data[k - 1] * b->data[L->col_ind->data[k - 1] - 1];
    }
  }
}

void crs_solve_utril_3args(const CRS_Matrix *L, emxArray_real_T *b, int offset)
{
  int n;
  int i;
  int b_i;
  int i1;
  int k;
  int i2;
  n = L->row_ptr->size[0];
  for (i = 0; i <= n - 2; i++) {
    b_i = L->row_ptr->data[i];
    i1 = L->row_ptr->data[i + 1] - 1;
    for (k = b_i; k <= i1; k++) {
      i2 = i + offset;
      b->data[i2] -= L->val->data[k - 1] * b->data[(L->col_ind->data[k - 1] +
        offset) - 1];
    }
  }
}

void crs_solve_utril_initialize(void)
{
}

void crs_solve_utril_terminate(void)
{
}

File: api/crs/crs_solve_utril.m
function b = crs_solve_utril(L, b, varargin)
% crs_solve_utril Solves L\b, where L is unit lower triangular
%     b = crs_solve_utril(L, b) solves L\b, where L stores the strict 
%     lower triangular part of L. The right-hand side vector will be 
%     overwritten by L\b at output.
%
%     b = crs_solve_utril(L, b, offset) uses b(offset+1:offset+n) 
%     instead of b for the rhs and solution vector, where n=size(L,1).
%
% See also crs_solve_tril, crs_solve_trilt

%#codegen -args {crs_matrix, coder.typeof(0, [inf,1])}
%#codegen crs_solve_utril_3args -args {crs_matrix, m2c_vec, int32(0)}

coder.inline('always');

if nargin < 3
    offset = int32(0);
else
    offset = int32(varargin{1});
end
n = int32(length(L.row_ptr) - 1);

for i = 1:n
    for k = L.row_ptr(i):L.row_ptr(i+1) - 1
        b(i+offset) = b(i+offset) - L.val(k) * b(L.col_ind(k)+offset);
    end
end

function test %#ok<DEFNU>
%!test
%! for k=1:100
%!     L = tril(rand(10,10), -1); b = rand(10,1);
%!     if condest(L) < 1.e5
%!         L = crs_createFromSparse(sparse(L));
%!         us = crs_solve_utril(L, b);
%!         assert(norm((L+speye(10)) * us - b) <= 1.e-10);
%!     end
%! end
